File: solutions/go/nucleotide-count/4/nucleotide_count.go
package dna

import (
	"errors"
	"fmt"
)

// Histogram is a mapping from nucleotide to its count in given DNA.
type Histogram map[rune]int

// DNA is a list of nucleotides.
type DNA []rune

var ns = []rune{'A', 'C', 'G', 'T'}

// Counts generates a histogram of valid nucleotides in the given DNA.
// Returns an error if d contains an invalid nucleotide.
func (d DNA) Counts() (Histogram, error) {
	h := make(Histogram)
	for _, n := range ns {
		h[n] = 0
	}
	for _, n := range d {
		_, ok := h[n]
		if !ok {
			return h, errors.New(fmt.Sprintf("nucleotides must be one of %v", ns))
		}
		h[n]++
	}
	return h, nil
}

File: solutions/zig/nucleotide-count/1/nucleotide_count.zig
pub const NucleotideError = error{Invalid};

pub const Counts = struct {
    a: u32,
    c: u32,
    g: u32,
    t: u32,
};

pub fn countNucleotides(s: []const u8) NucleotideError!Counts {
    var a: u32 = 0;
    var c: u32 = 0;
    var g: u32 = 0;
    var t: u32 = 0;
    for (s) |nucleotide| {
        switch (nucleotide) {
            'A' => a += 1,
            'C' => c += 1,
            'G' => g += 1,
            'T' => t += 1,
            else => return NucleotideError.Invalid,
        }
    }
    return Counts{ .a = a, .c = c, .g = g, .t = t };
}
